File: core/src/main/java/dev/vibris/core/ControlSession.java
package dev.vibris.core;

import dev.vibris.protocol.v1.ServerMessage;
import io.grpc.stub.StreamObserver;

final class ControlSession {
    private final StreamObserver<ServerMessage> responses;
    private String workspaceId = "";
    private String processInstanceId = "";
    private boolean connected = true;

    ControlSession(StreamObserver<ServerMessage> responses) {
        this.responses = responses;
    }

    synchronized void identify(String workspaceId, String processInstanceId) {
        this.workspaceId = workspaceId;
        this.processInstanceId = processInstanceId;
    }

    synchronized String workspaceId() {
        return workspaceId;
    }

    synchronized String processInstanceId() {
        return processInstanceId;
    }

    synchronized boolean connected() {
        return connected;
    }

    synchronized void send(ServerMessage message) {
        if (!connected) return;
        try {
            responses.onNext(message);
        } catch (RuntimeException exception) {
            connected = false;
        }
    }

    synchronized void complete() {
        if (!connected) return;
        connected = false;
        responses.onCompleted();
    }

    synchronized void disconnect() {
        connected = false;
    }
}

File: core/src/main/kotlin/dev/vibris/core/ControlSession.kt
package dev.vibris.core

import dev.vibris.protocol.v1.ServerMessage
import io.grpc.stub.StreamObserver

internal class ControlSession(private val responses: StreamObserver<ServerMessage>) {
    private var workspaceIdValue = ""
    private var processInstanceIdValue = ""
    private var connectedValue = true

    @Synchronized
    fun identify(workspaceId: String, processInstanceId: String) {
        workspaceIdValue = workspaceId
        processInstanceIdValue = processInstanceId
    }

    @Synchronized
    fun workspaceId(): String = workspaceIdValue

    @Synchronized
    fun processInstanceId(): String = processInstanceIdValue

    @Synchronized
    fun connected(): Boolean = connectedValue

    @Synchronized
    fun send(message: ServerMessage): Boolean {
        if (!connectedValue) {
            return false
        }
        return try {
            responses.onNext(message)
            true
        } catch (_: RuntimeException) {
            connectedValue = false
            false
        }
    }

    @Synchronized
    fun complete() {
        if (!connectedValue) {
            return
        }
        connectedValue = false
        responses.onCompleted()
    }

    @Synchronized
    fun disconnect() {
        connectedValue = false
    }
}